### DISCUSSION TITLE
[Experiment] ViewState opportunities suggestions

### DISCUSSION THREAD

**Discussion by tgrapperon** [pointfreeco/swift-composable-architecture#1225] [Ideas]
Hello!
I was wondering yesterday if we couldn't improve the `ViewState` story in TCA, that is small observable sub-states for `ViewStore`s to work with instead of whole feature's states.
I've made [this quick proof of concept](https://github.com/pointfreeco/swift-composable-architecture/compare/main...tgrapperon:swift-composable-architecture:viewstate-opportunities?expand=1) that observes each `KeyPath` access to the `ViewStore`'s state and compare them with what's available in the `State`. If `State` offers properties that are not used, it presents a runtime warning where it proposes a `ViewState` declaration.

For example, activating this by using `.lookForViewStateOpportunities()` in the `Animation` case study gives:
<p align="center">
<img width="556" alt="Runtime Warning proposing a `ViewState` for the `Animation` case study" src="https://user-images.githubusercontent.com/35562418/181027424-16e7597a-93f2-4a81-8591-407912c4c89c.png">
</p>
Likely overkill in this case, but the `ViewState` declaration is spot-on.

This is of course not a perfect solution:
- There are potentially many false positives caused by direct accesses to `viewStore.state` and its descendants that are missed and considered unaccessed.
- There are probably cases where the warning could be emitted before an escaped `ViewStore` accesses the properties. For example in modal/navigation contexts. I've added a small delay to the comparison to give it more time to unfold/evaluate all the body's hierarchy, like in `GeometryReader`s closures for example.
- This is not working yet with composed `KeyPath`, for example `viewStore.array.count`, but I feel it should be possible to match them with more work.
- It's not working well yet with property wrappers. I've envisaged several approaches for the whole feature, but finally opted to exploit `custom-dump` facilities around `KeyPath`s, because establishing the link between a `KeyPath` and [its property name is far from trivial](https://github.com/pointfreeco/swift-custom-dump/blob/main/Sources/CustomDump/Conformances/KeyPath.swift). This doesn't work well with property wrappers right now, but it could hopefully be improved for a general purpose. In this case, this experiment could be updated to support `@BindableState` properties.

This is also not working well yet when already using a `ViewState` (for example, the one it suggested 😬)!

The code was written in one go, in some hackathon/proof of concept style where I only wanted to see if it could work. There are of course many ways to refactor/reorganize it, or maybe rethink this at a deeper level. It is maybe possible to build the `ViewStore`'s `removeDuplicates` automatically, but I don't think it would be worth it performance-wise if it ever works.

I'm not pushing it farther yet because of these issues. The `ViewState` declaration suggestion seems very enticing though.
What do you think about it?

**Comment by stephencelis** [comment]
@tgrapperon Thanks for sharing! This looks like a lot of fun! Will try to take it for a spin soon :smile:
